man/rows.1
.TH "ROWS" "1" "2020-12-16" "-0.4.2-dev-eaa8b1ac-" "rows Manual"
.SH NAME
rows \- 
.SH SYNOPSIS
.B rows
[OPTIONS] COMMAND [ARGS]...
.SH OPTIONS
.TP
\fB\-\-http\-cache\fP BOOLEAN
.PP
.TP
\fB\-\-http\-cache\-path\fP TEXT
.PP
.TP
\fB\-\-version\fP
Show the version and exit.
.SH COMMANDS
.PP
\fBconvert\fP
  Convert table on `source` URI to...
  See \fBrows-convert(1)\fP for full documentation on the \fBconvert\fP command.
.PP
\fBjoin\fP
  Join tables from `source` URIs using `key(s)`...
  See \fBrows-join(1)\fP for full documentation on the \fBjoin\fP command.
.PP
\fBsum\fP
  Sum tables from `source` URIs and save into...
  See \fBrows-sum(1)\fP for full documentation on the \fBsum\fP command.
.PP
\fBprint\fP
  Print a table
  See \fBrows-print(1)\fP for full documentation on the \fBprint\fP command.
.PP
\fBquery\fP
  Query a table using SQL
  See \fBrows-query(1)\fP for full documentation on the \fBquery\fP command.
.PP
\fBschema\fP
  Identifies table schema
  See \fBrows-schema(1)\fP for full documentation on the \fBschema\fP command.
.PP
\fBcsv-to-sqlite\fP
  Convert one or more CSV files to SQLite
  See \fBrows-csv-to-sqlite(1)\fP for full documentation on the \fBcsv-to-sqlite\fP command.
.PP
\fBsqlite-to-csv\fP
  Convert a SQLite table into CSV
  See \fBrows-sqlite-to-csv(1)\fP for full documentation on the \fBsqlite-to-csv\fP command.
.PP
\fBpgimport\fP
  Import a CSV file into a PostgreSQL table
  See \fBrows-pgimport(1)\fP for full documentation on the \fBpgimport\fP command.
.PP
\fBpgexport\fP
  Export a PostgreSQL table into a CSV file
  See \fBrows-pgexport(1)\fP for full documentation on the \fBpgexport\fP command.
.PP
\fBpdf-to-text\fP
  Extract text from a PDF
  See \fBrows-pdf-to-text(1)\fP for full documentation on the \fBpdf-to-text\fP command.
.PP
\fBcsv-merge\fP
  Lazily merge CSVs (even if the schemas...
  See \fBrows-csv-merge(1)\fP for full documentation on the \fBcsv-merge\fP command.
.PP
\fBcsv-clean\fP
  Create a consistent and clean version of a...
  See \fBrows-csv-clean(1)\fP for full documentation on the \fBcsv-clean\fP command.
.PP
\fBcsv-row-count\fP
  Lazily count CSV rows
  See \fBrows-csv-row-count(1)\fP for full documentation on the \fBcsv-row-count\fP command.
.PP
\fBcsv-split\fP
  Split CSV into equal parts (by number of...
  See \fBrows-csv-split(1)\fP for full documentation on the \fBcsv-split\fP command.
.PP
\fBlist-sheets\fP
  List sheets
  See \fBrows-list-sheets(1)\fP for full documentation on the \fBlist-sheets\fP command.


CHANGELOG.md
# rows' Log of Changes

## Version `0.3.1`

**Released on: (under development)**


### Bug Fixes

- [#200](https://github.com/turicas/rows/pull/200) Unescape HTML when exporting
  (thanks to [@arloc](https://github.com/arloc))


## Version `0.3.0`

**Released on: 2016-09-02**

### Backwards Incompatible Changes

### Bug Fixes

- Return `None` on XLS blank cells;
- [#188](https://github.com/turicas/rows/issues/188) Change `sample_size` on
  encoding detection.


### Enhancements and Refactorings

- `rows.fields.detect_fields` will consider `BinaryField` if all the values are
  `str` (Python 2)/`bytes` (Python 3) and all other fields will work only with
  `unicode` (Python 2)/`str` (Python 3);
- Plugins HTML and XPath now uses a better way to return inner HTML (when
  `preserve_html=True`);
- [#189](https://github.com/turicas/rows/issues/189) Optimize `Table.__add__`.


### New Features

- Support for Python 3 (finally!);
- `rows.fields.BinaryField` now automatically uses base64 to encode/decode;
- Added `encoding` information to `rows.Table` metadata in text plugins;
- Added `sheet_name` information to `rows.Table` metadata in XLS and XLSX
  plugins;
- [#190](https://github.com/turicas/rows/issues/190) Add `query_args` to
  `import_from_sqlite`;
- [#177](https://github.com/turicas/rows/issues/177) Add `dialect` to
  `export_to_csv`.


## Version `0.2.1`

**Released on: 2016-08-10**

### Backwards Incompatible Changes

- `rows.utils.export_to_uri` signature is now like `rows.export_to_*` (first
  the `rows.Table` object, then the URI)
- Changed default table name in `import_from_sqlite` and `export_to_sqlite`
  (from `rows` and `rows_{number}` to `table{number}`)


### Bug Fixes

- [#170](https://github.com/turicas/rows/issues/170) (SQLite plugin) Error
  converting `int` and `float` when value is `None`.
- [#168](https://github.com/turicas/rows/issues/168) Use `Field.serialize`
  if does not know the field type (affecting: XLS, XLSX and SQLite plugins).
- [#167](https://github.com/turicas/rows/issues/167) Use more data to detect
  dialect, delimit the possible delimiters and fallback to excel if can't
  detect.
- [#176](https://github.com/turicas/rows/issues/176) Problem using quotes on
  CSV plugin.
- [#179](https://github.com/turicas/rows/issues/179) Fix double underscore
  problem on `rows.utils.slug`
- [#175](https://github.com/turicas/rows/issues/175) Fix `None`
  serialization/deserialization in all plugins (and also field types)
- [#172](https://github.com/turicas/rows/issues/172) Expose all tables in `rows
  query` for SQLite databases
- Fix `examples/cli/convert.sh` (missing `-`)
- Avoids SQL injection in table name


### Enhancements and Refactorings

- Refactor `rows.utils.import_from_uri`
- Encoding and file type are better detected on `rows.utils.import_from_uri`
- Added helper functions to `rows.utils` regarding encoding and file
  type/plugin detection
- There's a better description of plugin metadata (MIME types accepted) on
  `rows.utils` (should be refactored to be inside each plugin)
- Moved `slug` and `ipartition` functions to `rows.plugins.utils`
- Optimize `rows query` when using only one SQLite source


## Version `0.2.0`

**Released on: 2016-07-15**

### Backwards Incompatible Changes

- `rows.fields.UnicodeField` was renamed to `rows.fields.TextField`
- `rows.fields.BytesField` was renamed to `rows.fields.BinaryField`

### Bug Fixes

- Fix import errors on older versions of urllib3 and Python (thanks to
  [@jeanferri](https://github.com/jeanferri))
- [#156](https://github.com/turicas/rows/issues/156) `BoolField` should not
  accept "0" and "1" as possible values
- [#86](https://github.com/turicas/rows/issues/86) Fix `Content-Type` parsing
- Fix locale-related tests
- [#85](https://github.com/turicas/rows/issues/85) Fix `preserve_html` if
  `fields` is not provided
- Fix problem with big integers
- [#131](https://github.com/turicas/rows/issues/131) Fix problem when empty
  sample data
- Fix problem with `unicode` and `DateField`
- Fix `PercentField.serialize(None)`
- Fix bug with `Decimal` receiving `''`
- Fix bug in `PercentField.serialize(Decimal('0'))`
- Fix nested table behaviour on HTML plugin

### General Changes

- (EXPERIMENTAL) Add `rows.FlexibleTable` class (with help on tests from
  [@maurobaraildi](https://github.com/maurobaraldi))
- Lots of refactorings
- Add `rows.operations.transpose`
- Add `Table.__repr__`
- Renamte `rows.fields.UnicodeField` to `rows.fields.TextField` and
  `rows.fields.ByteField` to `rows.fields.BinaryField`
- Add a man page (thanks to [@kretcheu](https://github.com/kretcheu))
- [#40](https://github.com/turicas/rows/issues/40) The package is available on
  Debian!
- [#120](https://github.com/turicas/rows/issues/120) The package is available
  on Fedora!
- Add some examples
- [#138](https://github.com/turicas/rows/issues/138) Add
  `rows.fields.JSONField`
- [#146](https://github.com/turicas/rows/issues/146) Add
  `rows.fields.EmailField`
- Enhance encoding detection using
  [file-magic](https://pypi.python.org/pypi/file-magic) library
- [#160](https://github.com/turicas/rows/issues/160) Add
  support for column get/set/del in `rows.Table`

### Tests

- Fix "\r\n" on tests to work on Windows
- Enhance tests with `mock` to assure some functions are being called
- Improve some tests

### Plugins

- Add plugin JSON (thanks [@sxslex](https://github.com/sxslex))
- [#107](https://github.com/turicas/rows/issues/107) Add `import_from_txt`
- [#149](https://github.com/turicas/rows/issues/149) Add `import_from_xpath`
- (EXPERIMENTAL) Add `import_from_ods`
- (EXPERIMENTAL) Add `import_from_parquet`
- Add `import_from_sqlite` and `export_to_sqlite` (implemented by
  [@turicas](https://github.com/turicas) with help from
  [@infog](https://github.com/infog))
- Add `import_from_xlsx` and `export_to_xlsx` (thanks to
  [@RhenanBartels](https://github.com/turicas/RhenanBartels))
- Autodetect delimiter in CSV files

rows/rows.py
# coding: utf-8

# Copyright 2014 Álvaro Justen <https://github.com/turicas/rows/>
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

import logging
import types

from collections import defaultdict, OrderedDict
from itertools import chain

from .converters import TYPE_CONVERTERS, TYPES


LOG_FORMAT = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'

def _str_decode(element, codec):
    if isinstance(element, str):
        return element.decode(codec)
    else:
        return element

def get_type(value):
    key_type = type(value)
    if key_type == unicode:
        return str
    return key_type


class BaseTable(object):
    '''Base class for the really useful table classes'''

    def __init__(self, fields, log_filename=None, log_level=logging.INFO,
            log_format=LOG_FORMAT, input_encoding='utf-8'):
        self.input_encoding = input_encoding
        self._rows = []
        self.logger = None

        self.fields = fields
        self.types = {} # TODO: auto-initialize with 'str'?
        self._fields_not_converted = 0
        self._rows_not_converted = 0
        self.converters = TYPE_CONVERTERS
        if log_filename is not None:
            self.logger = logging.Logger(log_filename, level=log_level)
            handler = FileHandler(log_filename)
            formatter = Formatter(LOG_FORMAT)
            handler.setFormatter(formatter)
            self.logger.addHandler(handler)

    def rename_field(self, old_name, new_name):
        self.fields[self.fields.index(old_name)] = new_name
        self.types[new_name] = self.types[old_name]
        del self.types[old_name]

    def convert_row(self, row):
        encoding = self.input_encoding
        converters = self.converters
        converted = []
        error_message = ('Could not convert field value "{value}" '
                         '(field {field}, type {type})')
        not_converted = self._fields_not_converted
        for index, field in enumerate(self.fields):
            try:
                converted_value = \
                        converters[self.types[field]](row[index], encoding)
            except ValueError:
                converted_value = None
                self._fields_not_converted += 1
                if self.logger is not None:
                    data = {'field': field, 'value': repr(row[index]),
                            'type': self.types[field]}
                    self.logger.warning(error_message.format(**data))
            converted.append(converted_value)
        if self._fields_not_converted > not_converted:
            self._rows_not_converted += 1
        return converted

    def identify_data_types(self, sample_size, skip=None):
        """Create ``self.types``, a ``dict`` in which each key is a table
        header (from ``self.fields``) and value is a type in:
        ``(int, float, datetime.date, datetime.datetime, str)``.

        The types are identified trying to convert each column value to each
        type.
        """

        # TODO: maybe use <https://github.com/scraperwiki/scrumble/>
        # TODO: test this sample algorithm

        sample_rows = self._get_sample(sample_size)
        columns = zip(*sample_rows)
        converters = self.converters
        input_encoding = self.input_encoding
        for i, header in enumerate(self.fields):
            column_types = list(TYPES)
            cant_be = set()
            try:
                column = set(columns[i])
            except IndexError:
                self.types[header] = str
            else:
                types = list(set([type(value) for value in column]) -
                             set([type(None)]))
                if len(types) == 1 and types[0] not in (str, unicode):
                    # all rows have the same type (!= str, unicode)
                    identified_type = types[0]
                elif not [value for value in column if unicode(value).strip()]:
                    # all rows with an empty field -> str (can't identify)
                    identified_type = str
                else:
                    # ok, let's try to identify the type of this column by
                    # converting every value in the sample
                    for value in column:
                        if unicode(value).strip() == '' or value is None:
                            continue
                        for type_ in column_types:
                            try:
                                converters[type_](value, input_encoding)
                            except (ValueError, TypeError):
                                if type_ == str:
                                    raise ValueError(
                                            '{} cant be {}'.format(repr(value),
                                        type_))
                                cant_be.add(type_)

                    for removed_type in cant_be:
                        column_types.remove(removed_type)
                    identified_type = column_types[0] # priorities matter
                self.types[header] = identified_type

        self.samples_read = len(sample_rows)

    def filter(self, func):
        # TODO: may use (and create) Table.copy?
        filtered = Table(fields=self.fields)

        filtered._rows = []
        for row in self._rows:
            row_as_dict = dict(zip(self.fields, self.convert_row(row)))
            if func(row_as_dict):
                filtered._rows.append(row)

        filtered.input_encoding = self.input_encoding
        filtered.converters = self.converters.copy()
        filtered.types = self.types

        return filtered


class LazyTable(BaseTable):
    def __init__(self, fields, iterable, log_filename=None,
            log_level=logging.INFO, log_format=LOG_FORMAT):
        super(LazyTable, self).__init__(fields, log_filename, log_level,
                log_format)
        self._rows = iterable

    def __iter__(self):
        return self

    def next(self):
        return dict(zip(self.fields, self.convert_row(self._rows.next())))

    def _get_sample(self, sample_size):
        if sample_size is not None:
            sample = []
            for index, row in enumerate(self._rows, start=1):
                sample.append(row)
                if index == sample_size:
                    break
            self._rows = chain(sample, self._rows)
        else:
            sample = list(self._rows)
            self._rows = iter(sample)
        return sample


class Table(BaseTable):


    def _get_sample(self, sample_size):
        if sample_size is not None:
            return self._rows[:sample_size]
        else:
            return self._rows


    def append(self, row):
        row = self._prepare_to_append(row)
        self._rows.append(row)
        if not self.types:
            self.types = {key: get_type(value)
                    for key, value in zip(self.fields, row)}
            self.append = self._append


    def _append(self, row):
        row = self._prepare_to_append(row)
        row_types = [get_type(value) for value in row]
        table_types = [self.types[key] for key in self.fields]
        if row_types != table_types:
            raise ValueError('Incorrect types')
        self._rows.append(row)


    def _prepare_to_append(self, item):
        if isinstance(item, dict):
            row = []
            for column in self.fields:
                if column in item:
                    value = item[column]
                else:
                    value = None
                row.append(value)
        elif isinstance(item, (tuple, set)):
            row = list(item)
        elif isinstance(item, list):
            row = item
        else:
            raise ValueError()
        if len(row) != len(self.fields):
            raise ValueError()
        return [_str_decode(value, self.input_encoding) for value in row]


    def extend(self, items):
        """Append a lot of items.
        ``items`` should be a list of new rows, each row can be represented as
        ``list``, ``tuple`` or ``dict``.
        If one of the rows causes a ``ValueError`` (for example, because it has
        more or less elements than it should), then nothing will be appended to
        ``Table``.
        """
        new_items = []
        for item in items:
            new_items.append(self._prepare_to_append(item))
        for item in new_items:
            self.append(item)


    def __len__(self):
        """Returns the number of rows. Same as ``len(list)``."""
        return len(self._rows)


    def __setitem__(self, item, value):
        if isinstance(item, (str, unicode)):
            if item not in self.fields:
                self.append_column(item, value)
            columns = zip(*self._rows)
            if not columns or len(value) != len(self):
                raise ValueError()
            else:
                columns[self.fields.index(item)] = value
                self._rows = [list(x) for x in zip(*columns)]
        elif isinstance(item, int):
            self._rows[item] = self._prepare_to_append(value)
        elif isinstance(item, slice):
            self._rows[item] = [self._prepare_to_append(v) for v in value]
        else:
            raise ValueError()


    def __getitem__(self, item):
        if isinstance(item, (str, unicode)):
            if item not in self.fields:
                raise KeyError()
            columns = zip(*self._rows)
            if not columns:
                return []
            else:
                return list(columns[self.fields.index(item)])
        elif isinstance(item, (int, slice)):
            return dict(zip(self.fields, self.convert_row(self._rows[item])))
        else:
            raise ValueError()


    def __delitem__(self, item):
        if isinstance(item, (str, unicode)):
            columns = zip(*self._rows)
            header_index = self.fields.index(item)
            del columns[header_index]
            del self.fields[header_index]
            self._rows = [list(row) for row in zip(*columns)]
        elif isinstance(item, (int, slice)):
            del self._rows[item]
        else:
            raise ValueError()


    def count(self, row):
        """Returns how many rows are equal to ``row`` in ``Table``.
        Same as ``list.count``.
        """
        return self._rows.count(self._prepare_to_append(row))


    def index(self, x, i=None, j=None):
        """Returns the index of row ``x`` in table (starting from zero).
        Same as ``list.index``.
        """
        x = self._prepare_to_append(x)
        if i is None and j is None:
            return self._rows.index(x)
        elif j is None:
            return self._rows.index(x, i)
        else:
            return self._rows.index(x, i, j)


    def insert(self, index, row):
        """Insert ``row`` in the position ``index``. Same as ``list.insert``.
        ``row`` can be ``list``, ``tuple`` or ``dict``.
        """
        self._rows.insert(index, self._prepare_to_append(row))


    def pop(self, index=-1):
        """Removes and returns row in position ``index``. ``index`` defaults
        to -1. Same as ``list.pop``.
        """
        return self._rows.pop(index)


    def remove(self, row):
        """Removes first occurrence of ``row``. Raises ``ValueError`` if
        ``row`` is not found. Same as ``list.remove``.
        """
        self._rows.remove(self._prepare_to_append(row))


    def reverse(self):
        """Reverse the order of rows *in place* (does not return a new
        ``Table``, change the rows in this instance of ``Table``).
        Same as ``list.reverse``.
        """
        self._rows.reverse()


    def add_field(self, name, values, position=None, row_as_dict=False):
        """Append a field to the table

        If `posision` is None, the field is added to the end
        """

        if (type(values) != types.FunctionType and \
            len(values) != len(self)) or \
           name in self.fields:
            raise ValueError()
        if position is None:
            insert_header = lambda name: self.fields.append(name)
            insert_data = lambda row, value: row.append(value)
        else:
            insert_header = lambda name: self.fields.insert(position, name)
            insert_data = lambda row, value: row.insert(position, value)
        for index, row in enumerate(self):
            if type(values) == types.FunctionType:
                if row_as_dict:
                    value = values({header: row[index] \
                                    for index, header in \
                                        enumerate(self.fields)})
                else:
                    value = values(row)
            else:
                value = values[index]
            if not row_as_dict:
                insert_data(row, _str_decode(value, self.input_encoding))
            else:
                row[name] = _str_decode(value, self.input_encoding)
        insert_header(name)
        self.types[name] = str


    def remove_field(self, field):
        field_index = self.fields.index(field)
        for row in self._rows:
            row.pop(field_index)
        self.fields.pop(field_index)
        del self.types[field]


    def order_by(self, column, ordering='asc'):
        index = self.fields.index(column)
        if ordering.lower().startswith('desc'):
            sort_function = lambda x, y: cmp(y[index], x[index])
        else:
            sort_function = lambda x, y: cmp(x[index], y[index])
        self._rows.sort(sort_function)


    def transpose(self):
        everything = [self.fields] + self._rows
        new = zip(*everything)
        self.fields = new[0]
        self._rows = new[1:]
        self.types = {}
        self.identify_data_types(sample_size=None)


    def __radd__(self, other):
        if other == 0:
            return self
        raise ValueError()


    def __add__(self, other):
        if other == 0:
            return self

        if type(self) != type(other) or self.fields != other.fields or \
                self.types != other.types:
            raise ValueError('iTables are incompatible '
                    '(fields or its types are different)')
        table = Table(fields=self.fields)
        table._rows = self._rows + other._rows
        table.types = self.types
        return table


def join(keys, *tables):
    '''Merge a list of tables, using `keys` to group rows'''

    if isinstance(keys, (str, unicode)):
        keys = (keys, )

    data = defaultdict(OrderedDict)
    fields = []
    types = {}
    for table in tables:
        types.update(table.types)

        for field in table.fields:
            if field not in fields:
                fields.append(field)

        for row in table:
            row_key = tuple([row[key] for key in keys])
            data[row_key].update(row)

    merged = Table(fields=fields)
    merged.types = types
    merged._rows = []
    for row in data.values():
        new_row = []
        for field in fields:
            if field not in row:
                new_row.append(None)
            else:
                new_row.append(row[field])
        merged._rows.append(new_row)

    return merged
